File: ole.rb
class Ole < Sinatra::Base
  
  get '/ole' do
    ole
  end
  
  get '/ole/github' do
    payload = JSON.parse(params[:payload])
    last_commit = payload.commits.last.message
    repository = payload.repository.name
    ole_if_valid(repository, last_commit)
  end
  
  private
  
  def ole
    system("#{path_to_vlc} -I dummy --play-and-exit #{File.join(File.dirname(__FILE__), 'assets', 'ole_trumpet.mp3')}")
  end
  
  def path_to_vlc
    if RUBY_PLATFORM =~ /darwin/
      path = "/Applications/VLC.app/Contents/Mac OS/VLC"
      raise "VLC does not appear to be installed" unless File.exists?(path)
    else
      path = "vlc"
    end
    return path
  end
  
  def ole_if_valid(repo, msg)
    return unless ENV['ole_repos'].split(',').include?(repo) || msg =~ /\AOLE\:/
    ole
  end
end

File: ole.js
(function() {
  var exec, http, ole_path, path, server, vlc_path;
  http = require('http');
  exec = require('child_process').exec;
  path = require('path');
  server = http.createServer(function(request, response) {
    exec("" + (vlc_path()) + " -I dummy --play-and-exit " + (ole_path()), function(error, stdout, stderr) {
      if (stderr !== null) {
        console.log("VLC said:\n\n " + stderr);
      }
      if (error !== null) {
        return console.log(error);
      }
    });
    response.writeHead(200, {
      'Content-Type': 'text/plain'
    });
    return response.end("Ole'd!");
  });
  server.listen(9494);
  console.log("Ole server listening on port 9494!");
  ole_path = function() {
    return path.join(process.cwd(), 'assets', 'ole_trumpet.mp3');
  };
  vlc_path = function() {
    var vlc;
    if (require('os').type().match(/darwin/)) {
      vlc = {
        "/Applications/VLC.app/Contents/Mac OS/VLC": "vlc"
      };
    } else {
      vlc = "vlc";
    }
    return vlc;
  };
}).call(this);
